LKExp-mbed-x-nucleo-iks01a2/src/main.cpp

```diff
@@ -0,0 +1,191 @@
+  accelerometer->read_id(&id);
+  printf("LSM303AGR accelerometer           = 0x%X\r\n", id);
+  acc_gyro->read_id(&id);
+  printf("LSM6DSL accelerometer & gyroscope = 0x%X\r\n", id);
+    
+#ifdef STM32F769DISCO
+  BSP_LCD_Init();
+  BSP_LCD_LayerDefaultInit(0, LCD_FB_START_ADDRESS);
+  BSP_LCD_SelectLayer(0);
+  BSP_LCD_Clear(LCD_COLOR_WHITE);
+  
+  BSP_LCD_SetFont(&Font24);
+  BSP_LCD_DisplayStringAt(0, 0, (uint8_t*)"IMU example", CENTER_MODE);
+  uint8_t lcd_string[100];
+  BSP_LCD_SetFont(&Font20);
+#endif
+ 
+  while(1) {
+#ifndef STM32F769DISCO
+    printf("\r\n");
+#endif
+    
+    hum_temp->get_temperature(&value1);
+    hum_temp->get_humidity(&value2);
+#ifndef STM32F769DISCO
+    printf("HTS221: [temp] %7s C,   [hum] %s%%\r\n", print_double(buffer1, value1), print_double(buffer2, value2));
+#else
+    sprintf((char*)lcd_string,"HTS221: [temp] %7s C,   [hum] %s%%               ", print_double(buffer1, value1), print_double(buffer2, value2));
+    BSP_LCD_DisplayStringAt(0, 100, (uint8_t*)lcd_string, LEFT_MODE);
+#endif
+
+    press_temp->get_temperature(&value1);
+    press_temp->get_pressure(&value2);
+#ifndef STM32F769DISCO
+    printf("LPS22HB: [temp] %7s C, [press] %s mbar\r\n", print_double(buffer1, value1), print_double(buffer2, value2));
+    printf("---\r\n");
+#else
+    sprintf((char*)lcd_string,"LPS22HB: [temp] %7s C, [press] %s mbar               ", print_double(buffer1, value1), print_double(buffer2, value2));
+    BSP_LCD_DisplayStringAt(0, 150, (uint8_t*)lcd_string, LEFT_MODE);
+#endif
+    
+    magnetometer->get_m_axes(axes);
+#ifndef STM32F769DISCO
+    printf("LSM303AGR [mag/mgauss]:  %6ld, %6ld, %6ld\r\n", axes[0], axes[1], axes[2]);
+#else
+    sprintf((char*)lcd_string,"LSM303AGR [mag/mgauss]:  %6ld, %6ld, %6ld               ", axes[0], axes[1], axes[2]);
+    BSP_LCD_DisplayStringAt(0, 200, (uint8_t*)lcd_string, LEFT_MODE);
+#endif
+
+    accelerometer->get_x_axes(axes);
+#ifndef STM32F769DISCO
+    printf("LSM303AGR [acc/mg]:  %6ld, %6ld, %6ld\r\n", axes[0], axes[1], axes[2]);
+#else
+    sprintf((char*)lcd_string,"LSM303AGR [acc/mg]:  %6ld, %6ld, %6ld               ", axes[0], axes[1], axes[2]);
+    BSP_LCD_DisplayStringAt(0, 250, (uint8_t*)lcd_string, LEFT_MODE);
+#endif
+
+    acc_gyro->get_x_axes(axes);
+#ifndef STM32F769DISCO
+    printf("LSM6DSL [acc/mg]:      %6ld, %6ld, %6ld\r\n", axes[0], axes[1], axes[2]);
+#else
+    sprintf((char*)lcd_string,"LSM6DSL [acc/mg]: %6ld, %6ld, %6ld               ", axes[0], axes[1], axes[2]);
+    BSP_LCD_DisplayStringAt(0, 300, (uint8_t*)lcd_string, LEFT_MODE);
+#endif
+
+    acc_gyro->get_g_axes(axes);
+#ifndef STM32F769DISCO
+    printf("LSM6DSL [gyro/mdps]:   %6ld, %6ld, %6ld\r\n", axes[0], axes[1], axes[2]);
+#else
+    sprintf((char*)lcd_string,"LSM6DSL [gyro/mdps]: %6ld, %6ld, %6ld               ", axes[0], axes[1], axes[2]);
+    BSP_LCD_DisplayStringAt(0, 350, (uint8_t*)lcd_string, LEFT_MODE);
+#endif
+ 
+    if (mems_event) {
```

Comment:
pareil pour intégrer un peu de mbed et se faire la main, tu peux avoir ton event qui se retrouve dans un EventQueue.

> https://os.mbed.com/docs/mbed-os/v5.15/tutorials/the-eventqueue-api.html

Un exemple qui donne une idée (avec le serial mais c'est pareil) 

> https://gist.github.com/lonesometraveler/e3711f0f80046019e519cd84606d2618

L'idée est de supprimer `mems_event` et d'avoir le print dans un callback du event queue.

Reply:
Réalisé dans https://github.com/leka/LekaOS_Explorations/pull/2/commits/6972e90479fcb074e443ad5ce0e988a577b49813
J'ai pas réussi à le réaliser à travers ton exemple, j'en ai essayé d'autre avec [Simplify your code with mbed-events](https://os.mbed.com/blog/entry/Simplify-your-code-with-mbed-events/).

- Le premier exemple (Semaphore) ne convient pas car on ne retire pas l'équivalent du `mems_event`.
- Le second exemple (mbed-events) aurait pu fonctionner mais la fonction `attach` qui est directement relié à la fonction `rise` ou (`fall` dans l'exemple), et ne permettait que de prendre en argument un pointeur de fonction et non un `EventQueue`.
- Le troisième exemple (Manually deferring from ISR to a thread) a fonctionné sans soucis. Un `evenThread` a été ajouté pour que ça puisse fonctionner.